include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,630 @@
+           {
+              nbl::hlsl::uint32_t3 shifted = shift64ExtraRightJamming(zFrac0, zFrac1, zFrac2, -zExp);
+              zFrac0 = shifted.x;
+              zFrac1 = shifted.y;
+              zFrac2 = shifted.z;
+              zExp = 0;
+              
+              if (roundNearestEven)
+              {
+                 increment = zFrac2 < 0u;
+              }
+              else
+              {
+                 if (zSign != 0u)
+                 {
+                    increment = (FLOAT_ROUNDING_MODE == FLOAT_ROUND_DOWN) && (zFrac2 != 0u);
+                 }
+                 else
+                 {
+                    increment = (FLOAT_ROUNDING_MODE == FLOAT_ROUND_UP) && (zFrac2 != 0u);
+                 }
+              }
+           }
+        
+           if (increment)
+           {
+              nbl::hlsl::uint32_t2 added = add64(zFrac0, zFrac1, 0u, 1u);
+              zFrac0 = added.x;
+              zFrac1 = added.y;
+              zFrac1 &= ~((zFrac2 + uint32_t(zFrac2 == 0u)) & uint32_t(roundNearestEven));
+           }
+           else
+           {
+              zExp = LERP(zExp, 0, (zFrac0 | zFrac1) == 0u);
+           }
+           
+           return packFloat64(zSign, zExp, zFrac0, zFrac1);
+        }
+        
+        uint64_t normalizeRoundAndPackFloat64(uint32_t sign, int exp, uint32_t frac0, uint32_t frac1)
+        {
+           int shiftCount;
+           nbl::hlsl::uint32_t3 frac = nbl::hlsl::uint32_t3(frac0, frac1, 0u);
+        
+           if (frac.x == 0u)
+           {
+              exp -= 32;
+              frac.x = frac.y;
+              frac.y = 0u;
+           }
+        
+           shiftCount = countLeadingZeros32(frac.x) - 11;
+           if (0 <= shiftCount)
+           {
+              frac.xy = shortShift64Left(frac.x, frac.y, shiftCount);
+           }
+           else
+           {
+              frac.xyz = shift64ExtraRightJamming(frac.x, frac.y, 0u, -shiftCount);
+           }
+           exp -= shiftCount;
+           return roundAndPackFloat64(sign, exp, frac.x, frac.y, frac.z);
+        }
+        
+        static const uint64_t SIGN_MASK = 0x8000000000000000ull;
+        static const uint64_t EXP_MASK = 0x7FF0000000000000ull;
+        static const uint64_t MANTISA_MASK = 0x000FFFFFFFFFFFFFull;
+    }
+
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+
+        storage_t data;
+
+        // constructors
+        // TODO: specializations
+        template <typename T>
+        static emulated_float64_t create(T val)
+        {
+#ifndef __HLSL_VERSION
+            emulated_float64_t output;
+            output.data = reinterpret_cast<storage_t&>(val);
+            return output;
+#else
+            uint32_t lowBits;
+            uint32_t highBits;
+            asuint(val, lowBits, highBits);
+            
+            emulated_float64_t output;
+            output.data = (uint64_t(highBits) << 32) | uint64_t(lowBits);
+            return output;

```

Comment:
when does this even get used?

you obviously can't feed it a float64_t because that will emit the double SPIR-V capability (unless you optimize), and for `float32_t`, `float16_t` you'll get the bits in the wrong places.

And I'm not even gonna start on `T` being an integer

Reply:
removed